src/core/server/services/users/permissions/index.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  isSiteMember,
+  isSiteModerator,
+  PermissionsAction,
+} from "coral-common/permissions/types";
+import { MongoContext } from "coral-server/data/context";
+import { GQLUSER_ROLE } from "coral-server/graph/schema/__generated__/types";
+import {
+  pullUserMembershipScopes,
+  pullUserSiteModerationScopes,
+  updateUserRole,
+} from "coral-server/models/user";
+
+type SideEffect = (mongo: MongoContext, tenantID: string) => Promise<any>;
+
+export type PermissionsActionSideEffectTest = (
+  action: PermissionsAction
+) => SideEffect | undefined;
+
+const userShouldBeDemotedToCommenter: PermissionsActionSideEffectTest = ({
+  user,
+  newUserRole,
+  scopeDeletions,
+}) => {
+  if (
+    !scopeDeletions ||
+    (newUserRole !== GQLUSER_ROLE.MODERATOR &&
+      newUserRole !== GQLUSER_ROLE.MEMBER)
+  ) {
+    return;
+  }
+
+  const relevantScopes =
+    newUserRole === GQLUSER_ROLE.MODERATOR
+      ? user.moderationScopes
+      : user.membershipScopes;
+
+  const noneRemaining = !relevantScopes?.siteIDs?.find(
+    (siteID) => !scopeDeletions.includes(siteID)
+  );
+
+  if (noneRemaining) {
+    return (mongo: MongoContext, tenantID: string) =>
+      updateUserRole(mongo, tenantID, user.id, GQLUSER_ROLE.COMMENTER);
+  }
+
+  return async () => null;
+};
+
+const userShouldHaveModerationScopesRemoved: PermissionsActionSideEffectTest = ({
+  user,
+  newUserRole,
+}) => {
+  if (isSiteModerator(user) && newUserRole !== GQLUSER_ROLE.MODERATOR) {
+    return (mongo: MongoContext, tenantID: string) =>
+      pullUserSiteModerationScopes(
+        mongo,
+        tenantID,
+        user.id,
+        user.moderationScopes.siteIDs
+      );
+  }
+
+  return async () => null; // believe it or not it is yelling at me unless i do this
```

Comment:
Currently looking into this. It is complaining that "not all code paths return a value" even though the return type is potentially undefined